generation/generate-test.go

```diff
@@ -0,0 +1,335 @@
+	// get the json map of all known relations
+	b, err := os.ReadFile(followGraphPath)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// unpack into goland
+	var v map[string]map[string]interface{}
+	err = json.Unmarshal(b, &v)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	followMap := make(map[string][]string)
+	blockMap := make(map[string]map[string]bool)
+	for id, relations := range v {
+		var following []string
+		blockMap[id] = make(map[string]bool)
+		for relationId, status := range relations {
+			if isFollowing, ok := status.(bool); ok {
+				if isFollowing {
+					following = append(following, relationId)
+				} else {
+					blockMap[id][relationId] = true
+				}
+			}
+			followMap[id] = following
+		}
+	}
+	return followMap, blockMap, nil
+}
+
+func GetIdentities(fixturesRoot string) (map[string]string, error) {
+	b, err := os.ReadFile(path.Join(fixturesRoot, "secret-ids.json"))
+	if err != nil {
+		return nil, err
+	}
+	// secret-ids.json contains a map of ids -> {latest, folder}
+	var v map[string]map[string]interface{}
+	err = json.Unmarshal(b, &v)
+	if err != nil {
+		return nil, err
+	}
+	identities := make(map[string]string)
+	// however, we only want a mapping from id->foldername, so let's get that
+	for id, feedInfo := range v {
+		identities[id] = PickName(feedInfo)
+	}
+	return identities, nil
+}
+
+func getUniques(expectations map[string][]string) []string {
+	uniqueMap := make(map[string]bool)
+	uniques := make([]string, 0, len(expectations))
+
+	for id, replicatees := range expectations {
+		if _, ok := uniqueMap[id]; !ok && len(replicatees) > 0 {
+			uniques = append(uniques, id)
+			uniqueMap[id] = true
+		}
+		for _, replicateeId := range replicatees {
+			if _, ok := uniqueMap[replicateeId]; !ok {
+				uniques = append(uniques, replicateeId)
+				uniqueMap[replicateeId] = true
+			}
+		}
+	}
+	return uniques
+}
+
+// a couple useful helper functions :)
+func (g Generator) getNames(src []string) []string {
+	extractedNames := make([]string, 0, len(src))
+	for _, id := range src {
+		extractedNames = append(extractedNames, g.IDsToNames[id])
+	}
+	return extractedNames
+}
+
+func (g Generator) GetIDs(src []string) []string {
+	extractedIds := make([]string, 0, len(src))
+	for _, name := range src {
+		extractedIds = append(extractedIds, g.NamesToIDs[name])
+	}
+	return extractedIds
+}
+
+// TO DO:
+// * DONE		finish refactoring cli tool to use Generator struct instead of globals
+// * DONE		rewrite batch connect to use generator, and pass in a pair?
+// * think about how to pass in runtime args properly (document + export runtime args? try cryptix's functional pattern?)
+// * write tests to make senpai happy :^)
+// * fprintf to a slice or something, return the string from generateTest, then echo it in cli tool usecase
+// * some kind of workflow where we pass expectations as a slice of data? maybe not
+
+var writer io.Writer
```

Comment:
this global scares me.

Nasty things might happen if someone uses this package twice.

Reply:
I suggest doing this instead: ( The patch is here:  [no-global-patch.txt](https://github.com/ssb-ngi-pointer/netsim/files/6931094/no-global-patch.txt) )
```diff
diff --git a/generation/generate-test.go b/generation/generate-test.go
index 154023a..6689b82 100644
--- a/generation/generate-test.go
+++ b/generation/generate-test.go
@@ -28,6 +28,8 @@ type Generator struct {
 	currentlyExecuting map[string]bool
 	isBlocking         map[string]map[string]bool
 	Args               Args
+
+	output io.Writer
 }
 
 func check(err error) {
@@ -137,11 +139,12 @@ func (g Generator) GetIDs(src []string) []string {
 // * fprintf to a slice or something, return the string from generateTest, then echo it in cli tool usecase
 // * some kind of workflow where we pass expectations as a slice of data? maybe not
 
-var writer io.Writer
-
 func GenerateTest(args Args, expectations map[string][]string, outputWriter io.Writer) {
-	writer = outputWriter
-	g := Generator{Args: args, currentlyExecuting: make(map[string]bool)}
+	g := Generator{
+		Args:               args,
+		currentlyExecuting: make(map[string]bool),
+		output:             outputWriter,
+	}
 
 	var err error
 	// map of id -> [list of followed ids]
@@ -200,8 +203,8 @@ func GenerateTest(args Args, expectations map[string][]string, outputWriter io.W
 	// output `enter`, `load` stmts, sorted by puppet name
 	for _, puppetName := range puppetNames {
 		puppetId := g.NamesToIDs[puppetName]
-		fmt.Fprintf(writer, "enter %s\n", puppetName)
-		fmt.Fprintf(writer, "load %s %s\n", puppetName, puppetId)
+		fmt.Fprintf(g.output, "enter %s\n", puppetName)
+		fmt.Fprintf(g.output, "load %s %s\n", puppetName, puppetId)
 	}
 
 	// start the focus group
@@ -238,26 +241,26 @@ func (g Generator) batchConnect(p Pair) {
 
 func (g Generator) has(issuer string, names []string) {
 	for _, name := range names {
-		fmt.Fprintf(writer, "has %s %s@latest\n", issuer, name)
+		fmt.Fprintf(g.output, "has %s %s@latest\n", issuer, name)
 	}
 }
 
 func (g Generator) disconnect(issuer string, names []string) {
 	for _, name := range names {
-		fmt.Fprintf(writer, "disconnect %s %s\n", issuer, name)
+		fmt.Fprintf(g.output, "disconnect %s %s\n", issuer, name)
 	}
 }
 
 func (g Generator) connect(issuer string, names []string) {
 	for _, name := range names {
-		fmt.Fprintf(writer, "connect %s %s\n", issuer, name)
+		fmt.Fprintf(g.output, "connect %s %s\n", issuer, name)
 	}
 }
 
 func (g Generator) start(names []string) {
 	for _, name := range names {
 		if _, exists := g.currentlyExecuting[name]; !exists {
-			fmt.Fprintf(writer, "start %s %s\n", name, g.Args.SSBServer)
+			fmt.Fprintf(g.output, "start %s %s\n", name, g.Args.SSBServer)
 			g.currentlyExecuting[name] = true
 		}
 	}
@@ -277,7 +280,7 @@ func (g Generator) stop(names []string) {
 		}
 		if _, exists := g.currentlyExecuting[name]; exists {
 			delete(g.currentlyExecuting, name)
-			fmt.Fprintf(writer, "stop %s\n", name)
+			fmt.Fprintf(g.output, "stop %s\n", name)
 		}
 	}
 }
@@ -308,9 +311,9 @@ func (g Graph) RecurseFollows(id string, hopsLeft int, verbose bool) []Pair {
 		}
 		if verbose {
 			if hopsLeft == g.Gen.Args.MaxHops {
-				fmt.Fprintf(writer, "%d %s\n", g.Gen.Args.MaxHops-hopsLeft+1, g.Gen.IDsToNames[otherId])
+				fmt.Fprintf(g.Gen.output, "%d %s\n", g.Gen.Args.MaxHops-hopsLeft+1, g.Gen.IDsToNames[otherId])
 			} else {
-				fmt.Fprintf(writer, "%d %s (via %s)\n", g.Gen.Args.MaxHops-hopsLeft+1, g.Gen.IDsToNames[otherId], g.Gen.IDsToNames[id])
+				fmt.Fprintf(g.Gen.output, "%d %s (via %s)\n", g.Gen.Args.MaxHops-hopsLeft+1, g.Gen.IDsToNames[otherId], g.Gen.IDsToNames[id])
 			}
 		}
 		pairs = append(pairs, Pair{src: id, dst: otherId})
@@ -326,10 +329,10 @@ func (g Graph) RecurseFollows(id string, hopsLeft int, verbose bool) []Pair {
 
 func (g Generator) waitUntil(issuer string, names []string) {
 	for _, name := range names {
-		fmt.Fprintf(writer, "waituntil %s %s@latest\n", issuer, name)
+		fmt.Fprintf(g.Gen.output, "waituntil %s %s@latest\n", issuer, name)
 	}
 }
 
 func (g Generator) waitMs(ms int) {
-	fmt.Fprintf(writer, "wait %d\n", ms)
+	fmt.Fprintf(g.Gen.output, "wait %d\n", ms)
 }
```

Reply:
thanks for the patch!